pubsub/producer.go

```diff
@@ -0,0 +1,297 @@
+	p.promisesLock.Lock()
+	defer p.promisesLock.Unlock()
+	return len(p.promises)
+}
+
+// reproduce is used when Producer claims ownership on the pending
+// message that was sent to inactive consumer and reinserts it into the stream,
+// so that seamlessly return the answer in the same promise.
+func (p *Producer[Request, Response]) reproduce(ctx context.Context, value Request, oldKey string) (*containers.Promise[Response], error) {
+	id, err := p.client.XAdd(ctx, &redis.XAddArgs{
+		Stream: p.cfg.RedisStream,
+		Values: map[string]any{messageKey: p.mReq.Marshal(value)},
+	}).Result()
+	if err != nil {
+		return nil, fmt.Errorf("adding values to redis: %w", err)
+	}
+	p.promisesLock.Lock()
+	defer p.promisesLock.Unlock()
+	promise := p.promises[oldKey]
+	if oldKey == "" || promise == nil {
+		pr := containers.NewPromise[Response](nil)
+		promise = &pr
+	}
+	delete(p.promises, oldKey)
+	p.promises[id] = promise
+	return promise, nil
+}
+
+func (p *Producer[Request, Response]) Produce(ctx context.Context, value Request) (*containers.Promise[Response], error) {
+	p.once.Do(func() {
+		p.StopWaiter.CallIteratively(p.checkAndReproduce)
+		p.StopWaiter.CallIteratively(p.checkResponses)
+	})
+	return p.reproduce(ctx, value, "")
+}
+
+// Check if a consumer is with specified ID is alive.
+func (p *Producer[Request, Response]) isConsumerAlive(ctx context.Context, consumerID string) bool {
+	val, err := p.client.Get(ctx, heartBeatKey(consumerID)).Int64()
+	if err != nil {
+		return false
+	}
+	return time.Now().UnixMilli()-val < int64(p.cfg.KeepAliveTimeout.Milliseconds())
+}
+
+func (p *Producer[Request, Response]) havePromiseFor(messageID string) bool {
+	p.promisesLock.Lock()
+	defer p.promisesLock.Unlock()
+	_, found := p.promises[messageID]
+	return found
+}
+
+func (p *Producer[Request, Response]) checkPending(ctx context.Context) ([]*Message[Request], error) {
+	pendingMessages, err := p.client.XPendingExt(ctx, &redis.XPendingExtArgs{
+		Stream: p.cfg.RedisStream,
+		Group:  p.cfg.RedisGroup,
+		Start:  "-",
+		End:    "+",
+		Count:  100,
+	}).Result()
+
+	if err != nil && !errors.Is(err, redis.Nil) {
+		return nil, fmt.Errorf("querying pending messages: %w", err)
+	}
+	if len(pendingMessages) == 0 {
+		return nil, nil
+	}
+	// IDs of the pending messages with inactive consumers.
+	var ids []string
+	active := make(map[string]bool)
+	for _, msg := range pendingMessages {
+		// Ignore messages not produced by this producer.
+		if !p.havePromiseFor(msg.ID) {
+			continue
+		}
+		alive, found := active[msg.Consumer]
+		if !found {
+			alive = p.isConsumerAlive(ctx, msg.Consumer)
+			active[msg.Consumer] = alive
+		}
+		if alive {
+			continue
+		}
+		ids = append(ids, msg.ID)
+	}
+	if len(ids) == 0 {
+		log.Trace("There are no pending messages with inactive consumers")
+		return nil, nil
+	}
+	log.Info("Attempting to claim", "messages", ids)
+	claimedMsgs, err := p.client.XClaim(ctx, &redis.XClaimArgs{
```

Comment:
> That should fail if due to race a consumer XACKed one of the messages?

Yes, according to: https://redis.io/docs/latest/commands/xclaim/: "XCLAIM will not claim a message in the following cases: The message doesn't exist in the group PEL (i.e. it was never read by any consumer)"
(if the message was acked, it shouldn't be in PEL list anymore so xclaim should fail).

> If that succeeds, will XACK by the consumer fail?

I don't think so. If XClaim is followed by XAck by initial consumer, that should succeed. Nothing is preventing consumer ack_ing message claimed by other.


Now either way you are contradicting the assumption that once consumer becomes inactive, it can then become active with all the memory state it had. Added comment to the package explaining the assumption.